File: Day08/Part02/day08_02.java
import java.io.BufferedReader;
import java.io.FileReader;
import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;


public class day08_02 {


    public static void main(String[] args) {


        HashMap<String, Integer> registers = new HashMap<String, Integer>();
        Integer largestOverallRegister = Integer.MIN_VALUE;
        
        try {

            BufferedReader br = new BufferedReader(new FileReader("input.txt"));
            String line = br.readLine();

            while (line != null) {
                String[] splitLine = line.split("\\s+");
                // System.out.println(Arrays.toString(splitLine));

                // set registers to 0 if they don't exist
                registers.putIfAbsent(splitLine[0], 0);
                registers.putIfAbsent(splitLine[4], 0);

                // if evaluation is true, inc or dec the first register value
                if (evaluate(registers.get(splitLine[4]), splitLine[5], Integer.valueOf(splitLine[6]))) {
                    // System.out.print(String.valueOf(registers.get(splitLine[0])) + " " + splitLine[1] + " " + splitLine[2] + " ");
                    if (splitLine[1].equals("inc")) {
                        registers.put(splitLine[0], registers.get(splitLine[0]) + Integer.parseInt(splitLine[2]));
                    } 
                    if (splitLine[1].equals("dec")) {
                        registers.put(splitLine[0], registers.get(splitLine[0]) - Integer.parseInt(splitLine[2]));
                    }
                    // System.out.println(" = " + String.valueOf(registers.get(splitLine[0])));
                    if (registers.get(splitLine[0]) > largestOverallRegister) largestOverallRegister = registers.get(splitLine[0]);
                };

                // get next line
                line = br.readLine();
            }

            br.close();

        } catch (Exception e) {
            e.printStackTrace();
        }

        // find the largest register
        Integer largestCurrentRegister = Integer.MIN_VALUE;
        for (Map.Entry<String, Integer> entry : registers.entrySet()) {
            if (entry.getValue() > largestCurrentRegister) {
                largestCurrentRegister = entry.getValue();
            }
        }

        // print results
        System.out.println("Largest Register Overall: " + String.valueOf(largestOverallRegister));
        System.out.println("Largest Register Currently: " + String.valueOf(largestCurrentRegister));

    }


    private static boolean evaluate(Integer first, String comparison, Integer second) {
        // System.out.println(String.valueOf(first) + " " + comparison + " " + String.valueOf(second));
        boolean result = false;
        switch (comparison) {
            case "<":
                if (first < second) {
                    result = true;
                }
                break;
            case ">":
                if (first > second) {
                    result = true;
                }
                break;
            case "<=":
                if (first <= second) {
                    result = true;
                }
                break;
            case ">=":
                if (first >= second) {
                    result = true;
                }
                break;
            case "==":
                if (first.equals(second)) {
                    result = true;
                }
                break;
            case "!=":
                if (!first.equals(second)) {
                    result = true;
                }
                break;
            default:
                System.out.println("Error: Could not find: '" + comparison + "'");
                break;
        }

        // System.out.println("Result: " + result);
        return result;
    }


}

File: Day08/Part02/day08_02.py
class Instruction:
    
    def __init__(self, reg1, action, amount, reg2, comparison, comparison_value):
        self.reg1 = reg1
        self.action = action
        self.amount = amount
        self.reg2 = reg2 
        self.comparison = comparison
        self.comparison_value = comparison_value


def read_values(filename):
    file1 = open(filename, 'r')
    instructions = []
    next_line = file1.readline()
    while (next_line):
        split_line = next_line.split()
        instruction = Instruction(  split_line[0], 
                                    split_line[1], 
                                    int(split_line[2]), 
                                    split_line[4], 
                                    split_line[5],
                                    int(split_line[6]))
        instructions.append(instruction)
        next_line = file1.readline()
    return instructions


def set_up_registers(instructions):
    registers = {}
    for instruction in instructions:
        registers[instruction.reg1] = 0
    return registers


def perform_instruction(instruction, registers):
    # print(registers[instruction.reg2], instruction.comparison, instruction.comparison_value)
    if (eval(str(registers[instruction.reg2]) + instruction.comparison + str(instruction.comparison_value))):
        if (instruction.action == "inc"):
            registers[instruction.reg1] += instruction.amount
        else:
            registers[instruction.reg1] -= instruction.amount


if __name__ == "__main__":
    
    # first read all instructions
    instructions = read_values("input.txt")

    # then set all registers to initial value of 0
    registers = set_up_registers(instructions)

    # keep track of largest value
    largestEver = None

    # perform all the instructions
    for instruction in instructions:
        perform_instruction(instruction, registers)
        if (largestEver == None or registers[instruction.reg1] > largestEver):
            largestEver = registers[instruction.reg1]

    # find the largest register
    largestCurrent = None
    for k,v in registers.items():
        # print("Value:", v)
        if largestCurrent == None or v > largestCurrent:
            largestCurrent = v

    # print largest register values
    print("Largest Register Value Ever:", largestEver)
    print("Largest Register Value Currently:", largestCurrent)
